### PR TITLE
[topic-branch PR] Replace oldtagset with tagset

### PR DESCRIPTION
**Pull request by djmitche**
This is a rather literal replacement of the old tagset functionality (HashedTags, HashingTagsetAccumulator) with the new APIs in ./pkg/tagset.

The intention is that the diff should have high locality: each hunk should be reviewable on its own, more or less.  One notable exception to this is ./pkg/tagger/tagstore.EntityTags#computeCache, where the rewrite was a bit more substantial to use the SlicingBuilder, and involves storing a few more slices that were previously computed on the fly.

Subsequent commits to the topic branch will refactor to improve performance and address the TODO's.  You can see work-in-progress on those subsequent commits in https://github.com/DataDog/datadog-agent/compare/dustin.mitchell/pkg-tagset-oldtagset.

### CODE REVIEW

File: pkg/dogstatsd/server.go [hunk 1]
```diff
@@ -711,21 +710,17 @@ func (s *Server) storeMetricStats(sample metrics.MetricSample) {
...
-	key := s.Debug.keyGen.Generate(sample.Name, "", s.debugTagsAccumulator)
+	key := s.Debug.keyGen.Generate(sample.Name, "", tags)
 
 	// store
 	ms := s.Debug.Stats[key]
 	ms.Count++
 	ms.LastSeen = now
 	ms.Name = sample.Name
-	ms.Tags = strings.Join(s.debugTagsAccumulator.Get(), " ") // we don't want/need to share the underlying array
+	ms.Tags = tags.String()
```
Comment (ogaca-dd) [inline pkg/dogstatsd/server.go:723]: Note: Before  the separator was `" "` whereas now it is `", "`. As it is a debug function I don't think that is an issue.

File: pkg/tagger/tagstore/entity_tags.go [hunk 2]
```diff
@@ -43,35 +50,29 @@ func (e *EntityTags) getStandard() []string {
 }
 
 func (e *EntityTags) get(cardinality collectors.TagCardinality) []string {
-	return e.getHashedTags(cardinality).Get()
+	return e.getHashedTags(cardinality).UnsafeReadOnlySlice()
```
Comment (ogaca-dd) [inline pkg/tagger/tagstore/entity_tags.go:53]: Should you make a copy here as you did previously?

**Review by ogaca-dd — APPROVED** [review]
LGTM, great PR

File: pkg/aggregator/ckey/key.go [hunk 3]
```diff
@@ -33,32 +33,27 @@ type TagsKey uint64
...
-		hg: oldtagset.NewHashGenerator(),
-	}
+	return &KeyGenerator{}
 }
 
 // KeyGenerator generates hash for the given name, hostname and tags.
 // The tags don't have to be sorted and duplicated tags will be ignored while
 // generating the hash.
 // Not safe for concurrent usage.
 type KeyGenerator struct {
```
[the remark this answers is not in the document]
Comment (djmitche) [reply pkg/aggregator/ckey/key.go:43]: I will address that in a subsequent PR (https://github.com/DataDog/datadog-agent/commit/cdf54a775d7f135f41611e37f6b34c57260e8b00)

File: pkg/dogstatsd/server.go (hunk 1 above)
Comment (djmitche) [reply pkg/dogstatsd/server.go:723]: Oh, nicely spotted.  I agree that this is probably not an issue, now that you've noticed.

File: pkg/tagger/tagstore/entity_tags.go (hunk 2 above)
Comment (djmitche) [reply pkg/tagger/tagstore/entity_tags.go:53]: It doesn't hurt!

**Comment by djmitche** [comment]
Since this is merging to a topic branch, I'll merge once the fixes suggested above are green.

File: pkg/dogstatsd/server.go (hunk 1 above)
Comment (djmitche) [reply pkg/dogstatsd/server.go:723]: ..on the other hand, it's tested -- I'll fix it.

**Comment by djmitche** [comment]
hm, I think I need to rebase the topic branch!

**Pull request merged**